File: harvester-bot2.js
const CONSTANTS = require('./helper-package-js/constants');
const { ACTION, FACE, FLOWER, MOVE } = require('./helper-package-js/constants');

const { Action } = require('./helper-package-js/action');
const { Bot } = require('./helper-package-js/bot');
const { Position } = require('./helper-package-js/position');

class HarvesterBot2 extends Bot {
    constructor() {
        super("HarvesterBot2");

        this.spawnedHive = false;
    }

    getActions() {
        const actions = [];

        for (let beeCell of this.beeCells) {
            const bee = beeCell.bee;
            const pos = beeCell.pos;
            if (bee.pollen >= bee.count) {
                // If the bees have pollen, move to closest queen or hive.
                if (!this.isBesideHiveOrQueen(pos)) {
                    let minPath = this.getMinPath(pos, this.queenBee.pos);
                    for (let hiveCell of this.hiveCells) {
                        let path = this.getMinPath(pos, hiveCell.pos);
                        if (path.distance < minPath.distance) {
                            minPath = path;
                        }
                    }
                    actions.push(new Action({
                        type: ACTION.MOVE,
                        pos: pos,
                        move: minPath.move,
                        face: minPath.move - 1 /* Faces move direction. */,
                    }));
                }
            } else {
                // Move to closest flower to collect pollen.
                const firstFlower = this.flowerCells[0];
                let minPath = this.getMinPath(pos, firstFlower.pos);
                for (let i = 1; i < this.flowerCells.length; ++i) {
                    const path = this.getMinPath(pos, this.flowerCells[i].pos);
                    if (path.distance < minPath.distance) {
                        minPath = path;
                    }
                }
                actions.push(new Action({
                    type: ACTION.MOVE,
                    pos: pos,
                    move: minPath.move,
                    face: minPath.move - 1 /* Faces move direction. */,
                }));
            }
        }

        const queenPos = this.queenBee.pos;

        if (!this.spawnedHive && this.queenBee.pollen >= CONSTANTS.HIVE_POLLEN_AMOUNT) {
            actions.push(new Action({
                type: ACTION.CREATE_HIVE,
            }));
            this.spawnedHive = true;
        } else if (this.hiveCells.length > 0 && this.queenBee.pollen >= CONSTANTS.BEE_POLLEN_AMOUNT) {
            actions.push(new Action({
                type: ACTION.SPAWN,
                pos: this.hiveCells[0].pos,
                amount: this.queenBee.pollen / CONSTANTS.BEE_POLLEN_AMOUNT,
            }));
        }

        return actions;
    }
}

(async function main() {
    const harvesterBot2 = new HarvesterBot2();
    await harvesterBot2.run();
})();

File: harvester_bot2.py
from helper_package_python.constants import ActionType, BEE_POLLEN_AMOUNT

from helper_package_python.bot import Bot
from helper_package_python.action import Action

class HarvesterBot2(Bot):
	def __init__(self):
		super().__init__("HarvesterBot2")

	def getActions(self):
		actions = []

		for beeCell in self.beeCells:
			bee = beeCell.bee
			pos = beeCell.pos

			# If the bees have pollen, move to queen.
			if bee.pollen >= bee.count:
				if !self.isBesideHiveOrQueen(pos):
					minPath = self.getMinPath(pos, self.queenBee.pos)

					for hiveCell in self.hiveCells:
						path = self.getMinPath(pos, hiveCell.pos)
						if path.distance < minPath.distance:
							minPath = path

					actions.append(Action(
						ActionType.MOVE,
						pos,
						minPath.move,
						Face(minPath.move - 1) # Face move direction
					))
			else:
				# Move to closest flower to collect pollen.
				firstFlower = self.flowerCells[0]
				minPath = self.getMinPath(pos, firstFlower.pos)
				
				for i in range(1, len(self.flowerCells)):
					path = self.getMinPath(pos, self.flowerCells[i].pos)
					if path.distance < minPath.distance:
						minPath = path

				actions.append(Action(
					ActionType.MOVE,
					pos,
					minPath.move,
					Face(minPath.move - 1) # Face move direction
				))

		if not self.hiveCells:
			actions.append(Action(
				ActionType.CREATE_HIVE
			))
		else:
			actions.append(Action(
				ActionType.SPAWN,
				self.hiveCells[0].pos,
				self.queenBee.pollen / BEE_POLLEN_AMOUNT
			))

		return actions

def main():
	harvesterBot2 = HarvesterBot2()
	harvesterBot2.run()

main()